ajbsp_src/doc/ajbsp.6
.TH AJBSP 6 "June 2018"
.SH NAME
ajbsp \- a nodes builder for modern DOOM source ports
.SH SYNOPSIS
.B ajbsp
[\fI\,OPTION\/\fR]... \fI\,FILE\/\fR...
.SH DESCRIPTION
.B AJBSP
is a simple nodes builder for modern DOOM source ports.
It can build standard DOOM nodes, GL-Nodes, and XNOD format nodes.
Multiple wad files can be processed, and each file is modified in-place.
There is an option to backup files before processing them.
.SH OPTIONS
Short options begin with a single dash, followed by one
or more letters, where each letter is a distinct option
(i.e. short options can be mixed together).
Long options begin with two dashes, followed by a whole word.
When an option needs a value, it should be placed in the next
argument.
.PP
Filenames and options may be freely mixed, and the order of
options on the command line is not significant.
All files are processed with the same settings.
.PP
The special option '\-\-' causes all following arguments to be
interpreted as filenames.  This allows specifying a file which
begins with a hyphen.
.TP
.B \-v \-\-verbose
Produces more verbose output to the terminal.
Some warnings which are normally hidden (except
for a final tally) will be shown when enabled.
.TP
.B \-vv \-\-very\-verbose
Equivalent to using \-\-verbose twice, and causes lots of
wonderfully useless information about each level to be displayed.
.TP
.B \-vvv \-\-super\-verbose
Equivalent to using \-\-verbose three times, and enables the
display of all the minor issues (such as unclosed subsectors).
.TP
.B \-b \-\-backup
Backs up each input file before processing it.
The backup files will have the ".bak" extension
(replacing the ".wad" extension).  If the backup
file already exists, it will be silently overwritten.
.TP
.B \-f \-\-fast
Enables a faster method for selecting partition lines.
On large maps this can be significantly faster,
however the BSP tree may not be as good.
.TP
.B \-m \-\-map NAME(s)
Specifies one or more maps to process.
All other maps will be skipped (not touched at all).
The same set of maps applies to every given wad file.
The default behavior is to process every map in the wad.

Map names must be the lump name, like "MAP01" or "E2M3",
and cannot be abbreviated.  A range of maps can be
specified using a hyphen, such as "MAP04-MAP07".
Several map names and/or ranges can be given, using
commas to separate them, such as "MAP01,MAP03,MAP05".

NOTE: spaces cannot be used to separate map names.
.TP
.B \-n \-\-nogl
Disables building of GL-Nodes, only the normal nodes
are built.  Any existing GL-Nodes in a visited map
will be removed.
.TP
.B \-g \-\-gl5
Forces V5 format of GL-Nodes.  The normal behavior
is to build V2 format, and only switch to V5 format
when the level is too large (e.g. has too many segs).

Unless you are testing a source port, there is almost
no need to use this option.
.TP
.B \-x \-\-xnod
Forces XNOD (ZDoom extended) format of normal nodes.
Without this option, normal nodes will be built using
the standard DOOM format, and only switch to XNOD format
when the level is too large (e.g. has too many segs).

Using XNOD format can be better for source ports which
support it, since it provides higher accuracy for seg
splits.  However, it cannot be used with the original
DOOM.EXE or with Chocolate-Doom.
.TP
.B \-c \-\-cost ##
Sets the cost for making seg splits.
The value is a number between 1 and 32.
Larger values try to reduce the number of seg splits,
whereas smaller values produce more balanced BSP trees.
The default value is 11.

NOTE: this option has little effect when the \-\-fast
option is enabled.
.TP
.B \-o \-\-output FILE
This option is provided *only* for compatibility with
existing node builders.  It causes the input file to be
copied to the specified file, and that file is the one
processed.  This option cannot be used with multiple
input files, or with the \-\-backup option.
.TP
.B \-h \-\-help
Displays a brief help screen, then exits.
.TP
.B \-\-version
Displays a version string, then exits.
.SH EXIT STATUS
.TP
0
if OK.
.TP
1
if nothing was built (no matching maps).
.TP
2
if one or more maps failed to build properly.
.TP
3
if a fatal error occurred.
.SH COPYRIGHT
AJBSP is Copyright \(co 2018 Andrew Apted, Colin Reed, and
Lee Killough, et al.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.
.PP
This program is is distributed in the hope that it
will be useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.
.SH "SEE ALSO"
.PP
The AJBSP home page: https://gitlab.com/andwj/ajbsp


CHANGES.txt

CHANGES IN 7.70
===============

+  Heretic support!

-  Makefile.macos file, courtesy Doctor Nick

-  fixed error when Steepness setting is "NONE"
-  fixed blocked paths when using the "Alternate Starts" setting


CHANGES IN 7.666
================

+  UI: big revamp, two scrolling columns of modules
-  UI: main window can be vertically resized
-  UI: default save filename is the generated title name
-  UI: remember the previous folder when saving

-  UI: option to use mousewheel to change settings
-  UI: option to disable the overwrite-file warning
-  UI: the "Use" button in config manager now sets the seed too
-  UI: "@@" on the command line will forget current module

+  the "Weapons" button now controls how early or late new
   weapons are introduced into the wad

+  new "Park" rooms, naturalistic outdoor areas

+  hallways

+  more variety of window prefabs

+  ensure a secret exit can always be placed
-  fixed bug swapping normal-exit and secret-exit switches
-  support secret weapons (available earlier than usual)

+  add barrels, blood and corpses to rooms
-  fixed large teleport-in monsters getting stuck
-  support for the Quake3 map format in the CSG code
-  sky generator makes better looking mountains

-  new "Ramp Up" and "Bosses" buttons for monster panel
-  new themes: "Bit Mixed", "Original-ish", "Episode-ish"
-  new health/ammo settings: "Bit Less" and "Bit More"

!  removed Skulltag engine and modules
!  removed most other engines, since they did nothing at all
!  removed "Mode" button in UI, since it did very little
!  removed the non-standard player things (4001..4004)

-  got triple-key doors working again
-  got remote-activated doors working again
-  got lowering-bar fences working again

-  added exit signs to doors leading to exit room
-  renamed "Start Together" setting to "Alt-start Rooms"

-  improved title generator, removed pre-made backgrounds
-  support TGA format title pics when engine is GZDOOM

-  better scenic areas outside the map
-  raise outdoor fences when near an item pedestal (etc)
-  smooth corners where two fences meet

-  proper action system for remotely activated doors (etc)
-  introduced variation in how much symmetry to use
-  unused leaf rooms become "storage" for some minor items
-  lower tendency of creating long/thin rooms

-  various polishing of the shape rules
-  DOOM CSG: reworked how railing textures are handled
-  abbreviated config in OBLIGDAT lump and config manager
-  fixed issue with garbage characters in the Log Viewer



ajbsp_src/ajbsp.cc
//------------------------------------------------------------------------
//  MAIN PROGRAM
//------------------------------------------------------------------------
//
//  AJ-BSP  Copyright (C) 2001-2018  Andrew Apted
//          Copyright (C) 1994-1998  Colin Reed
//          Copyright (C) 1997-1998  Lee Killough
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//------------------------------------------------------------------------

#include "ajbsp.h"

#include <time.h>

#ifndef WIN32
#include <time.h>
#endif

namespace ajbsp {

#define MAX_SPLIT_COST  32


//
//  global variables
//

int opt_verbosity = 2;  // 0 is normal, 1+ is verbose

bool opt_fast		= false;

bool opt_no_gl		= false;
bool opt_force_v5	= false;
bool opt_force_xnod	= false;
int  opt_split_cost	= DEFAULT_FACTOR;

const char *Level_name;

map_format_e Level_format;

int total_built_maps = 0;
int total_failed_maps = 0;


typedef struct map_range_s
{
	const char *low;
	const char *high;

} map_range_t;

std::vector< map_range_t > map_list;


// this is > 0 when PrintMapName() is used and the current line
// has not been terminated with a new-line ('\n') character.
static int hanging_pos;

static void StopHanging()
{
	if (hanging_pos > 0)
	{
		hanging_pos = 0;

		printf("\n");
		fflush(stdout);
	}
}


//
//  show an error message and terminate the program
//
void FatalError(const char *fmt, ...)
{
	va_list arg_ptr;

	static char buffer[MSG_BUF_LEN];

	va_start(arg_ptr, fmt);
	vsnprintf(buffer, MSG_BUF_LEN-1, fmt, arg_ptr);
	va_end(arg_ptr);

	buffer[MSG_BUF_LEN-1] = 0;

	StopHanging();

	fprintf(stderr, "\nFATAL ERROR: %s", buffer);

	exit(3);
}


void PrintMsg(const char *fmt, ...)
{
	va_list arg_ptr;

	static char buffer[MSG_BUF_LEN];

	va_start(arg_ptr, fmt);
	vsnprintf(buffer, MSG_BUF_LEN-1, fmt, arg_ptr);
	va_end(arg_ptr);

	buffer[MSG_BUF_LEN-1] = 0;

	StopHanging();

	printf("%s", buffer);
	fflush(stdout);
}


void PrintVerbose(const char *fmt, ...)
{
	if (opt_verbosity < 1)
		return;

	va_list arg_ptr;

	static char buffer[MSG_BUF_LEN];

	va_start(arg_ptr, fmt);
	vsnprintf(buffer, MSG_BUF_LEN-1, fmt, arg_ptr);
	va_end(arg_ptr);

	buffer[MSG_BUF_LEN-1] = 0;

	StopHanging();

	printf("%s", buffer);
	fflush(stdout);
}


void PrintDetail(const char *fmt, ...)
{
	if (opt_verbosity < 2)
		return;

	va_list arg_ptr;

	static char buffer[MSG_BUF_LEN];

	va_start(arg_ptr, fmt);
	vsnprintf(buffer, MSG_BUF_LEN-1, fmt, arg_ptr);
	va_end(arg_ptr);

	buffer[MSG_BUF_LEN-1] = 0;

	StopHanging();

	printf("%s", buffer);
	fflush(stdout);
}


void PrintMapName(const char *name)
{
	if (opt_verbosity >= 1)
	{
		PrintMsg("  %s\n", name);
		return;
	}

	// display the map names across the terminal

	if (hanging_pos >= 68)
		StopHanging();

	printf("  %s", name);
	fflush(stdout);

	hanging_pos += strlen(name) + 2;
}


void DebugPrintf(const char *fmt, ...)
{
	(void) fmt;
}


//------------------------------------------------------------------------


static bool CheckMapInRange(const map_range_t *range, const char *name)
{
	if (strlen(name) != strlen(range->low))
		return false;

	if (strcmp(name, range->low) < 0)
		return false;

	if (strcmp(name, range->high) > 0)
		return false;

	return true;
}


static bool CheckMapInMaplist(short lev_idx)
{
	// when --map is not used, allow everything
	if (map_list.empty())
		return true;

	short lump_idx = edit_wad->LevelHeader(lev_idx);

	const char *name = edit_wad->GetLump(lump_idx)->Name();

	for (unsigned int i = 0 ; i < map_list.size() ; i++)
		if (CheckMapInRange(&map_list[i], name))
			return true;

	return false;
}


static build_result_e BuildFile()
{
	int num_levels = edit_wad->LevelCount();

	if (num_levels == 0)
	{
		PrintMsg("  No levels in wad\n");
		return BUILD_OK;
	}

	int visited = 0;
	int failures = 0;

	// prepare the build info struct
	nodebuildinfo_t nb_info;

	nb_info.factor		= opt_split_cost;
	nb_info.gl_nodes	= ! opt_no_gl;
	nb_info.fast		= opt_fast;

	nb_info.force_v5	= opt_force_v5;
	nb_info.force_xnod	= opt_force_xnod;

	build_result_e res = BUILD_OK;

	// loop over each level in the wad
	for (int n = 0 ; n < num_levels ; n++)
	{
		if (! CheckMapInMaplist(n))
			continue;

		visited += 1;

		if (n > 0 && opt_verbosity >= 2)
			PrintMsg("\n");

		res = AJBSP_BuildLevel(&nb_info, n);

		// handle a failed map (due to lump overflow)
		if (res == BUILD_LumpOverflow)
		{
			res = BUILD_OK;
			failures += 1;
			continue;
		}

		if (res != BUILD_OK)
			break;

		total_built_maps += 1;
	}

	StopHanging();

	if (res == BUILD_Cancelled)
		return res;

	if (visited == 0)
	{
		PrintMsg("  No matching levels\n");
		return BUILD_OK;
	}

	PrintMsg("\n");

	total_failed_maps += failures;

	if (res == BUILD_BadFile)
	{
		PrintMsg("  Corrupted wad or level detected.\n");
		return BUILD_OK;
	}

	if (failures > 0)
	{
		PrintMsg("  Failed maps: %d (out of %d)\n", failures, visited);
	}

	if (true)
	{
		PrintMsg("  Serious warnings: %d\n", nb_info.total_warnings);
	}

	if (opt_verbosity >= 1)
	{
		PrintMsg("  Minor issues: %d\n", nb_info.total_minor_issues);
	}

	return BUILD_OK;
}

void VisitFile(const char *filename)
{

	PrintMsg("\n");
	PrintMsg("Building %s\n", filename);

	edit_wad = Wad_file::Open(filename, 'a');
	if (! edit_wad)
		FatalError("Cannot open file: %s\n", filename);

	if (edit_wad->IsReadOnly())
	{
		delete edit_wad; edit_wad = NULL;

		FatalError("file is read only: %s\n", filename);
	}

	build_result_e res = BuildFile();

	// this closes the file
	delete edit_wad; edit_wad = NULL;

	if (res == BUILD_Cancelled)
		FatalError("CANCELLED\n");
}

static void ShowVersion()
{
	printf("ajbsp " AJBSP_VERSION "  (" __DATE__ ")\n");

	fflush(stdout);
}


static void ShowBanner()
{
	printf("+-----------------------------------------------+\n");
	printf("|   AJBSP " AJBSP_VERSION "   (C) 2018 Andrew Apted, et al   |\n");
	printf("+-----------------------------------------------+\n");

	fflush(stdout);
}


bool ValidateMapName(char *name)
{
	if (strlen(name) < 2 || strlen(name) > 8)
		return false;

	if (! isalpha(name[0]))
		return false;

	for (const char *p = name ; *p ; p++)
	{
		if (! (isalnum(*p) || *p == '_'))
			return false;
	}

	// Ok, convert to upper case
	for (char *s = name ; *s ; s++)
	{
		*s = toupper(*s);
	}

	return true;
}


void ParseMapRange(char *tok)
{
	char *low  = tok;
	char *high = tok;

	// look for '-' separator
	char *p = strchr(tok, '-');

	if (p)
	{
		*p++ = 0;

		high = p;
	}

	if (! ValidateMapName(low))
		FatalError("illegal map name: '%s'\n", low);

	if (! ValidateMapName(high))
		FatalError("illegal map name: '%s'\n", high);

	if (strlen(low) < strlen(high))
		FatalError("bad map range (%s shorter than %s)\n", low, high);

	if (strlen(low) > strlen(high))
		FatalError("bad map range (%s longer than %s)\n", low, high);

	if (low[0] != high[0])
		FatalError("bad map range (%s and %s start with different letters)\n", low, high);

	if (strcmp(low, high) > 0)
		FatalError("bad map range (wrong order, %s > %s)\n", low, high);

	// Ok

	map_range_t range;

	range.low  = low;
	range.high = high;

	map_list.push_back(range);
}


void ParseMapList(const char *from_arg)
{
	// create a mutable copy of the string
	// [ we will keep long-term pointers into this buffer ]
	char *buf = StringDup(from_arg);

	char *arg = buf;

	while (*arg)
	{
		if (*arg == ',')
			FatalError("bad map list (empty element)\n");

		// find next comma
		char *tok = arg;
		arg++;

		while (*arg && *arg != ',')
			arg++;

		if (*arg == ',')
		{
			*arg++ = 0;
		}

		ParseMapRange(tok);
	}
}

//
//  the program starts here
//
int main(const char *filename)
{
	// sanity check on type sizes (useful when porting)
	CheckTypeSizes();

	ShowVersion();

	ShowBanner();

	VisitFile(filename);

	PrintMsg("\n");

	if (total_built_maps == 0)
	{
		PrintMsg("NOTHING was built!\n");
		total_failed_maps = 0;
		return 1;
	}
	else
	{
		PrintMsg("Ok, built nodes for %d map%s!\n",
				total_built_maps, (total_built_maps == 1 ? "" : "s"));
		total_built_maps = 0;
		total_failed_maps = 0;
	}

	// that's all folks!
	return 0;
}

} //namespace ajbsp

int AJBSP_Build(const char *filename)
{
	return ajbsp::main(filename);
}

//--- editor settings ---
// vi:ts=4:sw=4:noexpandtab
